File: 450. Delete Node in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        TreeNdoe* pre = NULL;
        TreeNdoe* targetNode = search(root, pre, key);
        if (!targetNode) return root;
        TreeNode* left = targetNode->left, *right = targetNode->right;
        if (left && right) {

        } else if (left) {

        } else if (right) {

        } else {
            if (is_left)
                pre->left = NULL;
            else
                pre->right = NULL;
            delete targetNode;
        }
    }

    TreeNode* search(TreeNode* root, TreeNdoe* pre, int key) {
        if (!root) return NULL;
        if (root->val == key) return root;
        else if (root->val > key)
            return search(root->left, root, key);
        return search(root->right, root, key);
    }
};

File: go/src/typical_problems/二叉树/450. Delete Node in a BST/450. Delete Node in a BST.go
package main

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Val > key {
		root.Left = deleteNode(root.Left, key)
	} else if root.Val < key {
		root.Right = deleteNode(root.Right, key)
	} else {
		if root.Left == nil {
			root = root.Right
		} else if root.Right == nil {
			root = root.Left
		} else {
			// 寻找左子树的最大节点
			tmp := root.Left
			for tmp.Right != nil {
				tmp = tmp.Right
			}
			root.Val = tmp.Val
			root.Left = deleteNode(root.Left, tmp.Val)
		}
	}
	return root
}
